packages/foundry/contracts/hooks/rebalancer/Rebalancer.sol

```
@@ -0,0 +1,633 @@
+
+        uint256 length = amountsOut.length;
+        for (uint256 i = 0; i < length; i++) {
+            lpBptAmount[i] -= amountsOut[i];
+        }
+
+        bool hasRemovedAll = true;
+
+        for (uint256 i = 0; i < length; i++) {
+            if (lpBptAmount[i] != 0) {
+                hasRemovedAll = false;
+            }
+        }
+
+        isStillLp[pool][msg.sender] = hasRemovedAll;
+    }
+
+    /***************************************************************************
+                                  Hook Functions
+    ***************************************************************************/
+    ///@inheritdoc IHooks
+    function onRegister(
+        address,
+        address pool,
+        TokenConfig[] memory,
+        LiquidityManagement calldata
+    ) public override onlyVault returns (bool) {
+        ///@dev removing for testing
+        // if (!IBasePoolFactory(weightedPoolFactory).isPoolFromFactory(pool)) {
+        //     revert OnlyWeightedPoolsAllowed();
+        // }
+
+        emit RebalancerHookRegistered(address(this), pool);
+
+        return true;
+    }
+
+    /// @inheritdoc IHooks
+    function getHookFlags() public pure override returns (HookFlags memory) {
+        HookFlags memory hookFlags;
+        hookFlags.shouldCallBeforeAddLiquidity = true;
+        hookFlags.shouldCallAfterRemoveLiquidity = true;
+        hookFlags.shouldCallComputeDynamicSwapFee = true;
+        hookFlags.shouldCallAfterSwap = true;
+        return hookFlags;
+    }
+
+    /// @inheritdoc BaseHooks
+    function onBeforeAddLiquidity(
+        address router,
+        address,
+        AddLiquidityKind,
+        uint256[] memory,
+        uint256,
+        uint256[] memory,
+        bytes memory
+    ) public view override onlySelfRouter(router) returns (bool) {
+        // We only allow addLiquidity via the Router/Hook itself
+        return true;
+    }
+
+    /// @inheritdoc BaseHooks
+    function onAfterRemoveLiquidity(
```

Could also be a before hook, to fail faster?

---

I mean if I do it before hook I would need to do a rebalance i.e removing and adding liquidity before the swap wouldn't that change the swap qoute for that swap?